core/src/main/scala/cats/data/Func.scala

```diff
@@ -0,0 +1,128 @@
+package cats
+package data
+
+/**
+ * [[Func]] is a function `A => F[B]`.
+ *
+ * See: [[https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf The Essence of the Iterator Pattern]]
+ */
+sealed abstract class Func[F[_], A, B] { self =>
```

Comment:
Yes, the semantics of `product` and `compose` formed around `Func` is different from Kliesli composition, which is described in detail in https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf.